transformer_engine/jax/csrc/modules.cpp

```diff
@@ -425,9 +425,9 @@ void LayerNormBackwardImpl(size_t n, size_t hidden, bool zero_centered_gamma, fl
         (workspace_size + barrier_size + dgamma_part_size + dbeta_part_size);
 
     void *workspace = cublasLtMetaManager::Instance().GetWorkspace(total_workspace_size);
-    void *barrier = static_cast<char *>(workspace) + workspace_size;
-    void *dgamma_part = static_cast<char *>(barrier) + barrier_size;
+    void *dgamma_part = static_cast<char *>(workspace) + workspace_size;
     void *dbeta_part = static_cast<char *>(dgamma_part) + dgamma_part_size;
+    void *barrier = static_cast<char *>(dbeta_part) + dbeta_part_size;
```

Comment:
We could still have alignment issues if the hidden dim is not nice. A more robust approach is to pad the sizes to a multiple of 128:
```c++
auto pad_size = [](size_t size) -> size_t {
    constexpr size_t alignment = 128;
    return ((size + alignment - 1) / alignment) * alignment;
};
workspace_size = pad_size(workspace_size);
barrier_size = pad_size(barrier_size);
dgamma_part_size = pad_size(dgamma_part_size);
dbeta_part_size = pad_size(dbeta_part_size);
```
Then it doesn't matter what order we partition the allocated workspace buffer.

Reply:
I think the alignment issue will not appear after this change, because only the tuned kernels require 16 bytes memory alignment for the vector accesses, and the currently tuned kernels ensure the gamma and beta size be the multiplier of 16 bytes. However, I agree that we can manually pad those pointers to have better ensurence. I offline discussed with Ming about this issue last week and I will enhance the current allocator to make it support asking multiple 128 bytes aligned workspace at a time.

Reply:
Adding support to auto padding and mutliple sizes allocation.
Refer to [commits/7d681e57](https://github.com/NVIDIA/TransformerEngine/pull/379/commits/7d681e57102ad268584391c1ba3e377eaf14a07d0)